pkg/util/testing/wrappers.go

```diff
@@ -331,3 +333,286 @@ func (e *EventRecordWrapper) Message(message string) *EventRecordWrapper {
+func (b *ResourceClaimSpecBuilder) WithAdminAccess(enabled bool) *ResourceClaimSpecBuilder {
+	if len(b.spec.Devices.Requests) > 0 {
+		lastIdx := len(b.spec.Devices.Requests) - 1
+		if b.spec.Devices.Requests[lastIdx].Exactly != nil {
+			b.spec.Devices.Requests[lastIdx].Exactly.AdminAccess = ptr.To(enabled)
+		}
+	}
+	return b
+}
+
+// WithDeviceConstraints adds device constraints to the spec
+func (b *ResourceClaimSpecBuilder) WithDeviceConstraints(requestNames []string, matchAttribute string) *ResourceClaimSpecBuilder {
+	constraint := resourcev1.DeviceConstraint{
+		Requests:       requestNames,
+		MatchAttribute: ptr.To(resourcev1.FullyQualifiedName(matchAttribute)),
+	}
+	b.spec.Devices.Constraints = append(b.spec.Devices.Constraints, constraint)
+	return b
+}
+
+// WithDeviceConfig adds device configuration to the spec
+func (b *ResourceClaimSpecBuilder) WithDeviceConfig(requestName, driver string, parameters []byte) *ResourceClaimSpecBuilder {
+	config := resourcev1.DeviceClaimConfiguration{
+		Requests: []string{requestName},
+		DeviceConfiguration: resourcev1.DeviceConfiguration{
+			Opaque: &resourcev1.OpaqueDeviceConfiguration{
+				Driver:     driver,
+				Parameters: runtime.RawExtension{Raw: parameters},
+			},
+		},
+	}
+	b.spec.Devices.Config = append(b.spec.Devices.Config, config)
+	return b
+}
+
+// FirstAvailableRequest adds a FirstAvailable device request
+func (b *ResourceClaimSpecBuilder) FirstAvailableRequest(requestName, deviceClassName string) *ResourceClaimSpecBuilder {
+	req := resourcev1.DeviceRequest{
+		Name: requestName,
+		FirstAvailable: []resourcev1.DeviceSubRequest{{
+			Name:            "sub1",
+			DeviceClassName: deviceClassName,
+		}},
+	}
+	b.spec.Devices.Requests = append(b.spec.Devices.Requests, req)
+	return b
+}
+
+// Build returns the built ResourceClaimSpec
+func (b *ResourceClaimSpecBuilder) Build() resourcev1.ResourceClaimSpec {
+	return b.spec
+}
+
+// ResourceClaimTemplateWrapper wraps a resourcev1.ResourceClaimTemplate
+type ResourceClaimTemplateWrapper struct {
+	resourcev1.ResourceClaimTemplate
+}
+
+// MakeResourceClaimTemplate creates a ResourceClaimTemplateWrapper with basic metadata
+func MakeResourceClaimTemplate(name, namespace string) *ResourceClaimTemplateWrapper {
+	return &ResourceClaimTemplateWrapper{
+		resourcev1.ResourceClaimTemplate{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+			Spec: resourcev1.ResourceClaimTemplateSpec{
+				Spec: NewResourceClaimSpecBuilder().Build(),
+			},
+		},
+	}
+}
+
+// DeviceRequest adds a basic device request with the specified name and device class
+func (r *ResourceClaimTemplateWrapper) DeviceRequest(requestName, deviceClassName string, count int64) *ResourceClaimTemplateWrapper {
+	builder := NewResourceClaimSpecBuilder()
+	builder.spec = r.Spec.Spec
+	builder.DeviceRequest(requestName, deviceClassName, count)
+	r.Spec.Spec = builder.Build()
+	return r
```

Comment:
Here I'm honestly not sure, why do we go via interim builder rather just updating the state of the wrapper?

Reply:
OTOH, the power of builder (wrapper) pattern is that it allows chaining, eg `NewResourceClaimSpecBuilder(r.spec.Spec).DeviceRequest(...).Obj()` ?

Reply:
Yeah I agree we can collapse the 4 lines into one, +1